### DISCUSSION TITLE
Animation Blueprint Linking and Sub Anim Instances to change overlay

### DISCUSSION THREAD

**Discussion by UserProfileAccountName** [Sixze/ALS-Refactored#36] [Ideas]
Sub Anim Instances can help encapsulate the overlay animation logic and make it easier to add new overlays.
Animation Blueprint Linking can help in the transitions.

https://docs.unrealengine.com/4.26/en-US/AnimatingObjects/SkeletalMeshAnimation/AnimHowTo/LinkedAnimBP/

https://docs.unrealengine.com/4.26/en-US/AnimatingObjects/SkeletalMeshAnimation/AnimHowTo/SubAnimInstance/

**Comment by Sixze** [comment]
I considered these options, but they have several drawbacks that make them difficult to use:
- The need to manually transfer data from the parent animation blueprint to the linked one. There is no option like "automatically provide all data from the parent animation blueprint to the linked one", but there is a lot of data and copying can turn into a nightmare.
- It is not possible to set up a smooth transition from one linked animated blueprint to another.

**Comment by AchimTuran** [comment]
We use this in our ALS fork. But we had some issue with linked AnimBPs as Montages are not supported and it breaks the packaged game as well. I did some fixed in the engine to get it working.

> It is not possible to set up a smooth transition from one linked animated blueprint to another.

I don't know if you're aware of that we can use a blending time or play montages. We use Montages or if we don't have one the bleeding option inside a linked AnimBP.

> The need to manually transfer data from the parent animation blueprint to the linked one. There is no option like "automatically provide all data from the parent animation blueprint to the linked one", but there is a lot of data and copying can turn into a nightmare.

What we did is to copy the states that are required for the current existing overlay states or the locomotion stuff. It was not that much to get it working.

This was a devlog we did about our system:
https://youtu.be/iUCneTEr1NA

**Reply by AchimTuran** [reply]
Forgot to mention that we use a interface here to copy the states and a UAnimBP base class which can be used to implement overlay states.

**Reply by Sixze** [reply]
So, I tried to move the standing states into a linked animation instance and everything was fine.... until I noticed that animation montages were not playing inside linked animation instances 😞.

I found your [pull request](https://github.com/EpicGames/UnrealEngine/pull/8401) that fixes this, and I really hope that epics will notice it, otherwise it will be very sad.

**Comment by Sixze** [comment]
So guys I was wrong 😁, but now I finally managed to move all overlay-related stuff into separate linked animation instances, see e73710234c069a47f0a98f2285f96aadb4f92bc5.

Also I will not stop on overlay states, but will also try to move other stuff into linked animation instances too.
